lang/semgrep-grammars/src/semgrep-hack/grammar.js

```diff
@@ -13,23 +13,58 @@ const base_grammar = require('tree-sitter-hacklang/grammar');
 module.exports = grammar(base_grammar, {
   name: 'hack',
 
-  conflicts: ($, previous) => previous.concat([
-  ]),
+  conflicts: ($, previous) => previous.concat([]),
 
-  /*
-     Support for semgrep ellipsis ('...') and metavariables ('$FOO'),
-     if they're not already part of the base grammar.
-  */
   rules: {
-  /*
-    semgrep_ellipsis: $ => '...',
+    /*
+      Support for semgrep ellipsis ('...')
+    */
 
-    _expression: ($, previous) => {
+    // For now, we give precedence to native ellipsis (aka variadic modifiers).
+    ellipsis: ($) => prec(-1, '...'),
+    deep_ellipsis: ($) => prec(-1, seq('<...', $._expression, '...>')),
+
+    // By using an empty statement, we leverage a statement type that is strictly only used within
+    // $._statement and nowhere else.
+    empty_statement: ($, previous) => {
       return choice(
-        $.semgrep_ellipsis,
-        ...previous.members
+        previous,
+        // For conflict with `ellipsis;`
+        // we want it to be read as an expression with a semicolon
+        // instead of a statement
+        prec(-1, $.ellipsis) // statement ellipsis
+      );
+    },
+
+    _expression: ($, previous) => {
+      return choice(previous, $.ellipsis, $.deep_ellipsis);
```

Comment:
Why do semgrep extension files return the values? When I tried just setting, it worked the same as far as I can tell. Is it to just differentiate new rules vs changed rules to the reader?

Reply:
I'm not sure what you mean. Are you just being confused by the different ways to define anonymous functions?

```js
(x) => x
```
is the same as
```js
(x) => { return x; }
```
which is itself the same as
```js
function(x) { return x; }
```

When extending a `choice()` we can use two ways:

* The inline way, which works only for `choice()` (e.g. not `seq()`). It really extends the original list of choices:
```js
_expression: ($, previous) => {
  return choice(...previous.members, $.ellipsis, $.deep_ellipsis);
}
```

* The simpler way:
```js
_expression: ($, previous) => {
  return choice(previous, $.ellipsis, $.deep_ellipsis);
}
```
which is more compactly written as
```js
_expression: ($, previous) => choice(previous, $.ellipsis, $.deep_ellipsis)
```

The simpler way results in extra nesting, and you'll see that in the ocaml code. I don't know if there's a strong reason to use one or the other. I imitated [what was done for tree-sitter-typescript](https://github.com/tree-sitter/tree-sitter-typescript/blob/master/common/define-grammar.js), which heavily extends tree-sitter-javascript.

Reply:
This answers my question. I was just confused about the code style choices. In the [original file](https://github.com/returntocorp/ocaml-tree-sitter-semgrep/blob/926b452e6c3e819a27b8477254ae31937ba1ca16/lang/semgrep-grammars/src/semgrep-hack/grammar.js#L25-L30), one rule is written with the `(x) => x` syntax and one rule is written with the `(x) => { return x; }`. I didn't know if there was an implicit preference in certain cases.

Reply:
Yeah, don't assume I was clear with this syntax myself when I wrote or copied that code :stuck_out_tongue: I learn a lot by explaining these things and writing docs.

Reply:
I improved the template in this PR: https://github.com/returntocorp/ocaml-tree-sitter-semgrep/pull/245

Also, I turned the comments above into documentation: https://semgrep.dev/docs/contributing/adding-a-language/#extending-the-original-grammar-with-semgrep-syntax